lib/propolis/src/tasks.rs

```diff
@@ -412,3 +416,95 @@ impl Drop for CtrlHeld<'_> {
         }
     }
 }
+
+/// Holds a group of tokio task [task::JoinHandle]s to be later joined as a
+/// group when they have all concluded.
+pub struct TaskGroup(Mutex<Vec<task::JoinHandle<()>>>);
+impl TaskGroup {
+    pub fn new() -> Self {
+        Self(Mutex::new(Vec::new()))
+    }
+
+    /// Add to the group of contained tasks
+    pub fn extend<I>(&self, tasks: I)
+    where
+        I: Iterator<Item = task::JoinHandle<()>>,
+    {
+        let mut guard = self.0.lock().unwrap();
+        guard.extend(tasks);
+    }
+
+    /// Block until all held tasks have been joined, returning any resulting
+    /// [task::JoinError]s after doing so.
+    pub fn block_until_joined(&self) -> Option<Vec<task::JoinError>> {
+        let mut guard = self.0.lock().unwrap();
+        let workers = std::mem::replace(&mut *guard, Vec::new());
+        if workers.is_empty() {
+            return None;
+        }
+
+        let rt = tokio::runtime::Handle::current();
+        let errors = rt.block_on(async {
+            FuturesUnordered::from_iter(workers)
+                .filter_map(|res| futures::future::ready(res.err()))
+                .collect::<Vec<_>>()
+                .await
+        });
+        if errors.is_empty() {
+            None
+        } else {
+            Some(errors)
+        }
+    }
+}
+
+/// Convenience type alias for the [thread::JoinHandle::join()] error type
+pub type ThreadErr = Box<dyn Any + Send + 'static>;
+
+/// Holds a group of thread [thread::JoinHandle]s to be later joined as a group
+/// when they have all concluded.
+pub struct ThreadGroup(Mutex<Vec<thread::JoinHandle<()>>>);
+impl ThreadGroup {
+    pub fn new() -> Self {
+        Self(Mutex::new(Vec::new()))
+    }
+
+    /// Add to group of contained threads
+    ///
+    /// The first (if any) [std::io::Error] encountered while among the
+    /// `threads` items will determine the return from this function.  All
+    /// non-error [thread::JoinHandle]s will be added to the group even if one
+    /// or more items are an `Error`.
+    pub fn extend<I>(&self, threads: I) -> std::io::Result<()>
+    where
+        I: Iterator<Item = std::io::Result<thread::JoinHandle<()>>>,
+    {
+        let mut guard = self.0.lock().unwrap();
+        let mut status = Ok(());
+        for result in threads {
+            match result {
+                Err(e) => {
+                    // record the first error
+                    if status.is_ok() {
+                        status = Err(e);
+                    }
+                }
+                Ok(hdl) => guard.push(hdl),
+            }
+        }
+        status
+    }
+
+    /// Block until all contained [thread::JoinHandle]s have been joined,
+    /// returning any resulting [ThreadErr]s after doing so.
+    pub fn block_until_joined(&self) -> Option<Vec<ThreadErr>> {
```

Comment:
nitpick, take it or leave it: i'm a bit unconvinced that `Option<Vec<_>>` is really worth it here...why not just represent the "no errors" case with an empty `Vec`? this type introduces the possibility of multiple "empty" cases (`None` and `Some(vec![])`), which calling code might need to handle – although we will never currently return a `Some` with an empty vec, the caller still may need to deal with it. 

this isn't a huge deal, though, and if returning an option is useful when calling this, i'm fine with it.

Reply:
I kinda rather `Some(errors)`, and leaving it up to the callee to do the right thing (not return `Some([])`).  Considering how minimal usage of this facility is so far, the cost of changing it later is pretty low.